### DISCUSSION TITLE
Set CVL to Float voltage when cells are balanced

### DISCUSSION THREAD

**Discussion by carywin** [Louisvdw/dbus-serialbattery#453] [Ideas]
CVL is currently set to 3.45 V/cell which is a good absorption voltage for gently approaching 100 % SoC.

With the BMS balancing the cells during constant voltage mode we typically want absorption phase to terminate when current drops below a threshold or by timeout.

I propose terminating the absorption phase when the cells arrive in balance, say with a max cell delta of 5 to 10 mV, configurable. We can then switch the CVL to a lower float value (3.30 V/cell?) to prevent micro-cycling the pack at high SoC.

**Comment by redmoon2711** [comment]
Hi,

i have create my own charge algorithm based on victron and add the function to break absorption, when cell delta is lower 0.02V.
If the battery inbalance and one cell reached 3.5V, it dynamically lower the CVL to keep the battery under control. 

The functions are not completely tested yet! 
```
        def manage_charge_voltage_step(self):
        """
        Charge algorithm:
            1. Bulk:        The battery is charged at maximum charge current 
                            until the voltage increases to the configured absorption voltage.
            
            2. Absorption:  The battery is charged at the configured absorption voltage, 
                            with the charge current slowly decreasing as the battery approaches full charge.
                            
            3. Float:       The battery voltage is maintained at the configured float voltage to prevent discharge.
        
        
        nessesary parameter:
            self.charge_voltage_state = {"state": 0, "counter": 0}  to def __init__
            
            LINEAR_LIMITATION_ENABLE = False
            CVCM_ENABLE = True
            
            MAX_CELL_VOLTAGE = 3.45
            FLOAT_CELL_VOLTAGE = 3.35
            
            TAIL_CURRENT = None  or  0 - 5
            MAX_ABSOPTION_TIME_SEC = 60 * 60
            MIN_DELTA_CELL_VOLTAGE = 0.02
        """
        
        if CVCM_ENABLE:
            voltage_list = []
            
            # get cell voltages
            for i in range(self.cell_count):
                if i < min(len(self.cells), self.cell_count):
                    voltage_list.append(self.cells[i].voltage) 

            # Re-Bulk
            if sum(voltage_list) <= FLOAT_CELL_VOLTAGE * self.cell_count - 0.4 \
            and self.charge_voltage_state["state"] != 0:
                self.charge_voltage_state["state"] = 0
                self.charge_voltage_state["counter"] = 0

            # Absorption
            if self.charge_voltage_state["state"] == 1:
                control_v = MAX_CELL_VOLTAGE * self.cell_count
                                
                # count up
                self.charge_voltage_state["counter"] += 1
                
                # switch to Float if counter over ABSOPTION_TIME_SEC
                # or cell delta is lower MIN_DELTA_CELL_VOLTAGE
                if self.charge_voltage_state["counter"] >= ABSOPTION_TIME_SEC \
                or max(voltage_list) - min(voltage_list) <= MIN_DELTA_CELL_VOLTAGE:
                    self.charge_voltage_state["state"] = 2
                    self.charge_voltage_state["counter"] = 0
                
                # switch to Float if current lower then TAIL_CURRENT
                elif TAIL_CURRENT:
                    if self.current <= TAIL_CURRENT:
                        self.charge_voltage_state["state"] = 2
                        self.charge_voltage_state["counter"] = 0
            
            # Float
            elif self.charge_voltage_state["state"] == 2:
                control_v = FLOAT_CELL_VOLTAGE * self.cell_count
            
            # Bulk
            else:
                control_v = MAX_CELL_VOLTAGE * self.cell_count + 0.2
                
                # switch to absorption if voltage over absorption voltage
                if sum(voltage_list) >= MAX_CELL_VOLTAGE * self.cell_count:
                    self.charge_voltage_state["state"] = 1
            
            # lower charge voltage to protect cells with high cell voltage
            if max(voltage_list) > MAX_CELL_VOLTAGE + 0.05:
                self.control_voltage -= 0.05
            
            elif max(voltage_list) <= MAX_CELL_VOLTAGE or self.control_voltage > control_v:
                self.control_voltage = control_v
```

**Reply by mr-manuel** [reply]
I was thinking about the same. I just wanted to fix https://github.com/Louisvdw/dbus-serialbattery/issues/421 first and then look at this.

My idea was something like that: https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/d0ad3948d02a7988acdc22cf1fb4ec408935a9f6/etc/dbus-serialbattery/config.default.ini#L13-L55

I also made the config file more clear, since before different options wear spread across the file.

**Comment by carywin** [comment]
I've been experimenting with adjusting the DVCC charge voltage limit via MQTT & OpenHab and it works well when "Enable DC Grid Feed" is disabled, the battery is gently discharged (~30W) down to the float voltage level which takes a couple of hours (but usually the sun goes down first, so the battery starts supplying loads).
However if DC grid feed is enabled, the moment the voltage limit is lowered to float value, the inverter starts discharging the battery at max power and feeding that back to the grid. This is unwanted behaviour for me, so I don't think manipulating the charge voltage limit is the right way to go.
The next experiment I want to try is turning off the "Charge enable" function once the battery is balanced and letting it settle from there.

**Reply by arrow1800** [reply]
>  The next experiment I want to try is turning off the "Charge enable" function once the battery is balanced and letting it settle from there.

but how ? i am having the same issue. changing CVL is'nt the same as 'disabling charging all together'. which parameter do we need to disable charging completely ?

**Reply by carywin** [reply]
DVCC Max Charge Current - ModbusTCP register 2705 or `com.victronenergy.settings/Settings/SystemSetup/MaxChargeCurrent` on dbus

**Reply by arrow1800** [reply]
not sure if that would work. in my case Max Charge current is ignored when battery is almost full and the charge algorithm is in Constant voltage mode. hence the reason i am asking :)

**Comment by mr-manuel** [comment]
For completion. This was implemented in the driver.
